src/datasources/balances-api/valk-balances-api.service.ts

```diff
@@ -0,0 +1,134 @@
+      'balances.providers.valk.apiKey',
+    );
+    this.baseUri = this.configurationService.getOrThrow<string>(
+      'balances.providers.valk.baseUri',
+    );
+    this.defaultExpirationTimeInSeconds =
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.default',
+      );
+    this.defaultNotFoundExpirationTimeSeconds =
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.notFound.default',
+      );
+    this.chainsConfiguration = this.configurationService.getOrThrow<
+      Record<number, ChainAttributes>
+    >('balances.providers.valk.chains');
+  }
+
+  async getBalances(args: {
+    chainId: string;
+    safeAddress: string;
+    fiatCode: string;
+  }): Promise<Balance[]> {
+    try {
+      const cacheDir = CacheRouter.getValkBalancesCacheDir(args);
+      const chainName = this._getChainName(args.chainId);
+      const url = `${this.baseUri}/balances/token/${args.safeAddress}?chain=${chainName}`;
+      const valkBalances = await this.dataSource.get<ValkBalance[]>({
+        cacheDir,
+        url,
+        notFoundExpireTimeSeconds: this.defaultNotFoundExpirationTimeSeconds,
+        networkRequest: { headers: { Authorization: `${this.apiKey}` } },
+        expireTimeSeconds: this.defaultExpirationTimeInSeconds,
+      });
+      return this._mapBalances(valkBalances, args.fiatCode);
+    } catch (error) {
+      throw new DataSourceError(
+        `Error getting ${args.safeAddress} balances from provider: ${asError(error).message}}`,
+      );
+    }
+  }
+
+  async clearBalances(args: {
+    chainId: string;
+    safeAddress: string;
+  }): Promise<void> {
+    const key = CacheRouter.getValkBalancesCacheKey(args);
+    await this.cacheService.deleteByKey(key);
+  }
+
+  private _mapBalances(
+    valkBalances: ValkBalance[],
+    fiatCode: string,
+  ): Balance[] {
+    return valkBalances.map((valkBalance) => {
+      const price = valkBalance.prices[fiatCode.toUpperCase()] ?? null;
+      const balanceAmount = getNumberString(valkBalance.balance);
+      const fiatBalance = getNumberString(
+        (valkBalance.balance / Math.pow(10, valkBalance.decimals)) * price,
+      );
+      const fiatConversion = getNumberString(price);
+
+      // Valk returns a string representing the native coin (e.g.: 'eth') as
+      // token_address for native coins balances. An hex is returned for ERC20 tokens.
+      return {
+        ...(isHex(valkBalance.token_address)
+          ? {
+              tokenAddress: valkBalance.token_address,
+              token: {
+                name: valkBalance.name,
+                symbol: valkBalance.symbol,
+                decimals: valkBalance.decimals,
+                logoUri: valkBalance.logo ?? '',
+              },
+              balance: balanceAmount,
+            }
+          : {
+              tokenAddress: null,
+              token: null,
+              balance: balanceAmount,
+            }),
+        fiatBalance,
```

Comment:
You can move `balance` out of the ternary.

```suggestion
            }
          : {
              tokenAddress: null,
              token: null,
            }),
        balance: balanceAmount,
        fiatBalance,
```

---

The below is a nit and not necessary so I'll leave the decision up to you.

If you extracted the mapping, you could be more explicit with the return types, e.g. something like:

```ts
mapErc20Balance(valkBalance: ValkBalance): Erc20Balance
mapNativeBalance(valkBalance: ValkBalance): NativeBalance
mapValkBalance(valkBalance: ValkBalance): Balance {
  return {
    ...isAddress(valkBalance.token_address) ? mapErc20Balance(valkBalance) : mapNativeBalance(valkBalance)
    fiatBalance,
    fiatConversion
  }
}
```

Reply:
Good suggestion! I've adopted this mapping functions approach 🙂 in [9ec50af](https://github.com/safe-global/safe-client-gateway/pull/1062/commits/9ec50afe2fbb0b467535f4121b8191e53db9e10e)